kernel/console/src/lib.rs

```diff
@@ -3,156 +3,166 @@
-			LockableIo::<_, MutexIrqSafe<SerialPort>, _>::from(serial_port.clone()),
-		);
-
-		let _taskref = spawn::new_task_builder(console_entry, (new_console, receiver))
-			.name(alloc::format!("console_loop_{:?}", serial_port_address))
-			.spawn()?;
-	}
-
-	// Err("console_connection_detector task returned unexpectedly")
+            continue;
+        }
+
+        let serial_port = match get_serial_port(serial_port_address) {
+            Some(sp) => sp.clone(),
+            _ => {
+                error!(
+                    "Serial port {:?} was not initialized, skipping console connection request",
+                    serial_port_address
+                );
+                continue;
+            }
+        };
+
+        let (sender, receiver) = async_channel::new_channel(16);
+        if serial_port.lock().set_data_sender(sender).is_err() {
+            warn!(
+                "Serial port {:?} already had a data sender, skipping console connection request",
+                serial_port_address
+            );
+            continue;
+        }
+
+        let _ = spawn::new_task_builder(shell_loop, (serial_port, serial_port_address, receiver))
+            .name(format!("{:?}_manager", serial_port_address))
+            .spawn()
+            .unwrap();
+    }
 }
 
+// TODO: Remove unwraps
+
+fn shell_loop(
+    (port, address, receiver): (
+        Arc<MutexIrqSafe<SerialPort>>,
+        SerialPortAddress,
+        Receiver<DataChunk>,
+    ),
+) {
+    info!("creating new tty for serial port {:?}", address);
+
+    let tty = tty::Tty::new();
+
+    let reader_task = spawn::new_task_builder(tty_to_port_loop, (port.clone(), tty.master()))
+        .name(format!("tty_to_{:?}", address))
+        .spawn()
+        .unwrap();
+    let writer_task = spawn::new_task_builder(port_to_tty_loop, (receiver, tty.master()))
+        .name(format!("{:?}_to_tty", address))
+        .spawn()
+        .unwrap();
+
+    let new_app_ns = mod_mgmt::create_application_namespace(None).unwrap();
+
+    // TODO: Replace shell with shell_2
+    let (app_file, _ns) =
+        mod_mgmt::CrateNamespace::get_crate_object_file_starting_with(&new_app_ns, "shell_2-")
+            .expect("Couldn't find shell in default app namespace");
+
+    let path = path::Path::new(app_file.lock().get_absolute_path());
+    let task = spawn::new_application_task_builder(path, Some(new_app_ns))
+        .unwrap()
+        .name(format!("{:?}_shell", address))
+        .block()
+        .spawn()
+        .unwrap();
+
+    let id = task.id;
+    let stream = Arc::new(tty.slave());
+    app_io::insert_child_streams(
+        id,
+        app_io::IoStreams {
+            discipline: Some(stream.discipline()),
+            stdin: stream.clone(),
+            stdout: stream.clone(),
+            stderr: stream,
+        },
+    );
+
+    task.unblock();
+    task.join().unwrap();
+
+    reader_task.kill(KillReason::Requested).unwrap();
+    writer_task.kill(KillReason::Requested).unwrap();
```

Comment:
I'm not sure how we should do this since they're blocked on a channel.

1. Spin in a loop using `try_receive` and checking an `AtomicBool`. This is terrible as it means `tty_to_port_loop` and `port_to_tty_loop` are always `Runnable`.
2. 
```rust
let mut data = [0; 256];
loop {
    let len = master.read(&mut data).unwrap();
    if should_exit.load(Ordering::SeqCst) {
        break;
    }
    port.lock().write(&data[..len]).unwrap();
}
```
and then write something to the buffer from `shell_loop` after setting `should_exit` to `true` to ensure it is tested. I'm not sure how we would do this for `port_to_tty_loop` as it's waiting on a serial port. Could we expose a function from `serial_port` that returns a write handle so that we can imitate input from the serial port?